### PR TITLE
[Design] Removing ComplexModelDTOModelBinder

### PR DESCRIPTION
**Pull request by harshgMSFT**
Not much design here other than simplification resulting from removal of complex model dto model binder. 
Reasoning: Highly coupled, very hard to extend, hard to understand and maintain due to the flow. 
Also used to model bind an unrelated model `ComplexDTOModel` which caused to break the parent-child relationship of model binding contexts.

### CODE REVIEW

**Comment by harshgMSFT** [comment]
cc: @dougbu , @rynowak

**Comment by rynowak** [comment]
I like the idea of it, I really don't like the idea of making changes here until we've got confidence that we're enforcing the right behavior with our tests. For one we lose the ability to do comparisons between our code and the original.

**Comment by rynowak** [comment]
Would be good to also review https://aspnet.codeplex.com/downloads/get/110748 for any interesting tidbits.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs [hunk 1]
```diff
@@ -236,24 +231,37 @@ private static bool CanUpdateReadOnlyProperty(Type propertyType)
...
+                    bindingContext,
+                    propertyModelName,
+                    propertyMetadata);
 
-            childContext.Model = dto;
+                var modelBindingResult =
+                    await bindingContext.OperationBindingContext.ModelBinder.BindModelAsync(propertyBindingContext);
+                if (modelBindingResult == null)
+                {
+                    // Could not bind. Add a result so MutableObjectModelBinder will check for [DefaultValue].
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: nit: you're in `MutableObjectModelBinder` :smiley_cat: think you want to refer to `ProcessResult()`

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs [hunk 2]
```diff
@@ -236,24 +231,37 @@ private static bool CanUpdateReadOnlyProperty(Type propertyType)
             return true;
         }
 
-        private async Task<ModelBindingResult> CreateAndPopulateDto(
+        private async Task<IDictionary<ModelMetadata, ModelBindingResult>> BindProperties(
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: `BindPropertiesAsync`

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs [hunk 3]
```diff
@@ -404,7 +414,7 @@ internal void ProcessDto(ModelBindingContext bindingContext, ComplexModelDto dto
 
             // For each property that ComplexModelDtoModelBinder attempted to bind, call the setter, recording
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: Comment no longer makes sense.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs [hunk 4]
```diff
@@ -404,7 +414,7 @@ internal void ProcessDto(ModelBindingContext bindingContext, ComplexModelDto dto
...
             // exceptions as necessary.
-            foreach (var entry in dto.Results)
+            foreach (var entry in result)
             {
                 var dtoResult = entry.Value;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: Variables probably need to be renamed.

**Comment by pranavkm** [comment]
Seems sound otherwise.

**Comment by dougbu** [comment]
@harshgMSFT suggest closing this PR

**Comment by harshgMSFT** [comment]
Closing.. will create a new one if we decide to fix this.

**Pull request closed without merging**